### ISSUE TITLE
Potential bug affecting the “BubbleGun bchains –fasta” command?

### ISSUE THREAD

**Issue by diaspj** [fawaz-dabbaghieh/bubble_gun#5]
Dear BubbleGun team,

I believe that there is a bug in the “BubbleGun bchains” command when the option “--fasta” is used. 

Bellow, I illustrate this potential bug with a specific example, using the information comprised in three different files: the BubbleGun .json file, the BubbleGun .fasta file and the vg VCF file generated using the “vg deconstruct” command over the pangenome original gfa file.

________________________________________________________________________________________

{"1": {"chain_id": 1, "ends": ["229577", "229575"], "bubbles": [{"type": "simple", "id": 1, "ends": ["229577", "229575"], "inside": ["394857", "229576"]}]}, "2": {"chain_id": 2, "ends": ["188813", "188811"], "bubbles": [{"type": "insertion", "id": 2, "ends": ["188813", "188811"], "inside": ["188812"]}]}, "3": {"chain_id": 3, "ends": ["51303", "51301"], "bubbles": [{"type": "simple", "id": 3, "ends": ["51303", "51301"], "inside": ["249385", "51302"]}]}, "4": {"chain_id": 4, "ends": ["39745", "39743"], "bubbles": [{"type": "simple", "id": 4, "ends": ["39745", "39743"], "inside": ["252991", "39744"]}]}, "5": {"chain_id": 5, "ends": ["237317", "237315"], "bubbles": [{"type": "simple", "id": 5, "ends": ["237317", "237315"], "inside": ["237316", "372104"]}]}, "6": {"chain_id": 6, "ends": ["361661", "361659"], "bubbles": [{"type": "simple", "id": 6, "ends": ["361661", "361659"], "inside": ["673904", "361660"]}]}, "7": {"chain_id": 7, "ends": ["376968", "376966"], "bubbles": [{"type": "simple", "id": 7, "

>bubble_1_allele_1_chain_1
GCA
>bubble_1_allele_2_chain_1
GTA
>bubble_2_allele_1_chain_2
AAA
>bubble_2_allele_2_chain_2
ACA
>bubble_4_allele_1_chain_3
CTGT
>bubble_4_allele_2_chain_3
CAGT
>bubble_3_allele_1_chain_3
CCC
>bubble_3_allele_2_chain_3
CTC
>bubble_5_allele_1_chain_4
GTAAT
>bubble_5_allele_2_chain_4
GTACT
>bubble_6_allele_1_chain_5
GTCCC
>bubble_6_allele_2_chain_5
GCTAC

Lactococcus_lactis_subsp._lactis_275_CP016699.1	62930	>229575>229577	G	C	60	.	AC=0;AF=0;AN=0;AT=>229575>229576>229577,>229575<394857>229577;NS=0;LV=0	GT

Lactococcus_lactis_subsp._lactis_229_CP016698.1	24872	>188811>188813	GT	G	60	.	AC=0;AF=0;AN=0;AT=>188811>188812>188813,>188811>188813;NS=0;LV=0	GT

Lactococcus_lactis_CBA3619_CP042409.1	53045	>51301>51303	T	C	60	.	AC=0;AF=0;AN=0;AT=>51301>51302>51303,>51301<249385>51303;NS=0;LV=0	GT

Lactococcus_lactis_CBA3619_CP042409.1	2051	>39743>39745	CT	GG	60	.	AC=0;AF=0;AN=0;AT=>39743>39744>39745,>39743<252991>39745;NS=0;LV=0	GT

Lactococcus_lactis_subsp._lactis_275_CP016700.1	48190	>237315>237317	ACTC	TTGA	60	.	AC=0;AF=0;AN=0;AT=>237315>237316>237317,>237315<372104>237317;NS=0;LV=0	GT
________________________________________________________________________________________

I also took a look at the BubbleGun code associated with the writing of the bchains fasta file:

________________________________________________________________________________________
from BubbleGun.path_extractor import sequence_extractor

def write_bubbles(graph, output_path):
    out_file = open(output_path, "w")

    for chain in graph.b_chains:
        for bubble in chain.bubbles:
            if not bubble.is_simple():
                continue

            first_path = [bubble.source.id, bubble.inside[0].id, bubble.sink.id]
            seq1 = sequence_extractor(graph, first_path)
            second_path = [bubble.source.id, bubble.inside[1].id, bubble.sink.id]
            seq2 = sequence_extractor(graph, second_path)

            read_name1 = ">bubble_" + str(bubble.id) + "_allele_1" + "_chain_" + str(chain.id)
            read_name2 = ">bubble_" + str(bubble.id) + "_allele_2" + "_chain_" + str(chain.id)

            out_file.write(read_name1 + "\n")
            out_file.write(seq1 + "\n")
            out_file.write(read_name2 + "\n")
            out_file.write(seq2 + "\n")

    out_file.close()
________________________________________________________________________________________

1)	First, based on the code reproduced above, it seems that the bubbles tagged with "insertion" or “super” will not be written. So, the "chain_id": 2 that comprises the bubble with "id": 2 should not be written in the BubbleGun fasta file, right?

2)	Second, the "chain_id": 3 comprises only a single bubble, and it is a simple bubble, identified as "id": 3. However, besides the expected alleles of chain_id 3

“>bubble_3_allele_1_chain_3” and “>bubble_3_allele_2_chain_3”

, there are also two additional alleles 

“>bubble_4_allele_1_chain_3” and “>bubble_4_allele_1_chain_3”

However, there is no bubble "id": 4 inside "chain_id": 3 in the json file.

3)	Third, after the above described behaviour, "chain_id": 4, which comprises only a single bubble identified as "id": 4, in the BubbleGun fasta file shows two alleles identified with the prefix “bubble_5”.

“>bubble_5_allele_1_chain_4” and “>bubble_5_allele_2_chain_4”

However, "id": 5 is associated with "chain_id": 5, not with "chain_id": 4. 

4)	Fourth, taking into the BubbleGun --fasta behaviour, which seems unconsistent with the output comprised in the BubbleGun .json file, trying to understand the nucleotides associated with each variant in the vg VCF file is a bit difficult, although it seems these also do not match

I hope that the BubbleGun team can shed some light on this issue, and confirm (or dismiss) my conclusion that there is bug in the “BubbleGun bchains –fasta” command.

Best regards, 

Paulo Dias

**Comment by fawaz-dabbaghieh** [comment]
Hi, thanks for raising an issue!

I still haven't processed the complete post, but in order to get this done quickly, could you provide the graph you're using and the commands?

Best,
Fawaz

**Comment by fawaz-dabbaghieh** [comment]
Yes, after reading your questions/concerns, there seems to be a bug, thanks for pointing this out. I have to investigate this further.

At first glance from these outputs, the issue could be a numbering issue (numbering for the IDs), that the IDs outputted in the JSON do not correspond well to the IDs in the fast file (which is an issue).

I'll try to look at this over the weekend.

Best,
Fawaz

**Comment by fawaz-dabbaghieh** [comment]
I just pushed a fix for this but, at least I think it fixes it now and I tested it on some small graphs I have.

To answer your questions:
1- Yes, only simple bubbles are written out
For questions 2, 3, and 4, I need the input file to know what's happening, I can't debug it from the output alone.

For the last question, I am not sure what VG deconstruct does internally and I haven't used it before. Thereofre, it's hard for me to match its output with mine without more information on the input file and what are your expectations for an output.
the --fasta option outputs two paths for each bubble which includes the source and sink nodes of the bubble, so it's not only outputting the inside nodes of the bubble, i.e. the two fasta sequences include the "variant" somewhere in the sequence flanked from both sides by the sequences of the source and sink node, if your graph looks like this
           -----G-----
ATG---                ----CGT
           -----T------
The output alleles will be
ATGGCGT
ATGTCGT

I will close the issue now hoping that my commit fixed it, if not, please feel free to open again with an input file I can use for debugging :)

Feel free of course to ask any more questions you have

Best,
Fawaz

**Comment by diaspj** [comment]
Dear Fawaz,

I have installed the BubbleGun with the fix, however, I think that your fix did not removed the bug at the level of "BubbleGun bchains --fasta" command.

I exemplify with the following two examples:



EXAMPLE: "chain_id" 1, bubble "id" 1
________________________________________________________________________________________________________________________________________________________________

Output of .json file:

{"1": {"chain_id": 1, "ends": ["229577", "229575"], "bubbles": [{"type": "simple", "id": 1, "ends": ["229577", "229575"], "inside": ["394857", "229576"]}]},



Output of vg deconstruct:

Lactococcus_lactis_subsp._lactis_275_CP016699.1	62930	>229575>229577	G	C	60	.	AC=0;AF=0;AN=0;AT=>229575>229576>229577,>229575<394857>229577;NS=0;LV=0	GT



Output of chainout command:

"BubbleGun -g Lactococcus_lactis_PANGENOME_Anchorwave.gfa chainout --json_file BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.json --chain_ids 1 --output_chain output_chain_1.gfa"

S	229577	A	LN:i:1	
L	229577	-	229576	-	0M
L	229577	-	394857	+	0M
S	394857	G	LN:i:1	
L	394857	-	229577	+	0M
L	394857	+	229575	-	0M
S	229576	G	LN:i:1	
L	229576	-	229575	-	0M
L	229576	+	229577	+	0M
S	229575	C	LN:i:1	
L	229575	+	229576	+	0M
L	229575	+	394857	-	0M



Output of .fasta file:

>bubble_1_allele_1_chain_1
GCG
>bubble_1_allele_2_chain_1
GTG



CONCLUSION: the output of all files are consistent except the output of the .fasta file. 

Considering the sequence on the .fasta file and of the .gfa file, maybe the bug in the .fasta program consists on S: 394857 and S: 229576 being considered as the "ends" nodes and the S: 229575 and S: 229577 as the "inside" nodes, with the following (wrong) bubble configuration:



		"229575" (bubble_1_allele_1_chain_1)
		_________ 
                    
 		    C 


    G		      	   	    G
________			________

"394857"			"229576"   

                    T (complementary of A, since in the vcf file the node <394857 is in the complementary DNA chain)		
		_________

		"229577" (bubble_1_allele_2_chain_1)




The proper (true) bubble configuration is: 


		"394857" (bubble_1_allele_1_chain_1)
		_________ 
                    
 		    C (complementary of G, since in the vcf file the node <394857 is in the complementary DNA chain)


    A		      	   	    C
________			________

"229577"			"229575"   

                    G 		
		_________

		"229576" (bubble_1_allele_2_chain_1)
________________________________________________________________________________________________________________________________________________________________







EXAMPLE: "chain_id" 2, bubble "id" 2
________________________________________________________________________________________________________________________________________________________________

Output of .json file:

"2": {"chain_id": 2, "ends": ["188813", "188811"], "bubbles": [{"type": "insertion", "id": 2, "ends": ["188813", "188811"], "inside": ["188812"]}]},



Output of vg deconstruct:

Lactococcus_lactis_subsp._lactis_229_CP016698.1	24872	>188811>188813	GT	G	60	.	AC=0;AF=0;AN=0;AT=>188811>188812>188813,>188811>188813;NS=0;LV=0	GT



Output of chainout command:

"BubbleGun -g Lactococcus_lactis_PANGENOME_Anchorwave.gfa chainout --json_file BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.json --chain_ids 2 --output_chain output_chain_2.gfa"

S	188812	T	LN:i:1	
L	188812	-	188811	-	0M
L	188812	+	188813	+	0M
S	188811	G	LN:i:1	
L	188811	+	188812	+	0M
L	188811	+	188813	+	0M
S	188813	A	LN:i:1	
L	188813	-	188811	-	0M
L	188813	-	188812	-	0M




Output of .fasta file:

>bubble_2_allele_1_chain_2
AAAGA
>bubble_2_allele_2_chain_2
AAATA



CONCLUSION: the output of all files are consistent except the output of the .fasta file. 

However, the bubbles tagged with "insertion" and "super" should not be in the output comprised in the .fasta file. Because "chain_id" 2, bubble "id" 2 is tagged with "insertion", from this point on, it is difficult to understand the behaviour of the BubbleGun software...

I think that the BubbleGun software might be outputting a sequence for the "insertion" since on the right part of the sequence we observe a "G", a "T" and a "A", which are the nucleotides that are present in the "output_chain_2.gfa" file, but this can be only a coincidence (and I can't explain the presence of those three "As", "AAA" on the left part of the sequence), so this bug  has to be identified by the developer of the software at the code level.
_________________________________________________________________________________________________________________________________________________________________


In order to attempt solving this problem, I am going to share the "pangenome .gfa file", the "vg deconstruct" VCF file and the BubbleGun .json, .gfa and .fasta files.

However, because this work is at the writing stage and still haven't been published, I will not share these files at the open github website, I am going to send the files directly to your e-mail.

Best regards, 

Paulo Dias

**Comment by fawaz-dabbaghieh** [comment]
Dear Paulo,

I am not able to replicate the results you are getting with my BubbleGun locally, I am using the latest version that I pushed recently and the results I am getting are consistent between the JSON and the FASTA files.

I am not checking the VCF because I cannot comment on what VG is doing exactly for their deconstruction and I would stick to the consistency issue between the JSON and the FASTA file inside BubbleGun.

I ran the following command on the smaller graph just for ease of debugging:
`BubbleGun -g BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.gfa bchains --bubble_json Lactococcus_lactis_PANGENOME_Anchorwave.json --fasta Lactococcus_lactis_PANGENOME_Anchorwave.fasta`

Looking at the first two chains in the JSON output from above, here is what it looks like
```
$ cat Lactococcus_lactis_PANGENOME_Anchorwave.json | jq | head -34

{
  "1": {
    "chain_id": 1,
    "ends": [
      "232029",
      "232025"
    ],
    "bubbles": [
      {
        "type": "simple",
        "id": 1,
        "ends": [
          "232027",
          "232029"
        ],
        "inside": [
          "232028",
          "397244"
        ]
      },
      {
        "type": "simple",
        "id": 2,
        "ends": [
          "232027",
          "232025"
        ],
        "inside": [
          "232026",
          "397245"
        ]
      }
    ]
  },
  "2": {
    "chain_id": 2,
    "ends": [
      "639144",
      "639146"
    ],
    "bubbles": [
      {
        "type": "simple",
        "id": 3,
        "ends": [
          "639144",
          "639146"
        ],
        "inside": [
          "639145",
          "690106"
        ]
      }
    ]
  },
```

So, chain 1 has two simple bubbles with IDs 1 and 2, and chain 2 has 1 simple bubble with ID 3. The reported FASTA looks like this at the beginning for those two chains
```
$ head -12 Lactococcus_lactis_PANGENOME_Anchorwave.fasta

>bubble_1_allele_1_chain_1
AACCT
>bubble_1_allele_2_chain_1
AATTT
>bubble_2_allele_1_chain_1
TTCCCC
>bubble_2_allele_2_chain_1
TTGTTC
>bubble_3_allele_1_chain_2
TTAC
>bubble_3_allele_2_chain_2
TAGC
```

Looking at the sequence of these nodes in the GFA file
```
S	232027	AA
S	232029	T
S	397244	AA
S	232028	CC
```
which is also consistent with the FASTA file, and the reason that the second allele is reverse complement is because of the edges, the edge between 232027 and 232028 is + +, and between 232028 and 232029 is also + +, so the allele is spelled in the forward direction (AACCT), and for other allele the edges are + -, so the reverse complement is taken which results in AATTT.

Moreover, I have looked at some of the bubbles reported as "insertion" in the JSON file I got from my run above, and they are not present in the FASTA file outputted.

Therefore, I can only assume that your BubbleGun version is different than mine. Make sure you installed the latest version after my edits, or delete completely the old version from your system and install it again. You can install with `python3 setup.py install --user` or `pip install -e .` when you're inside the bubble_gun directory that is cloned from github.

Best,
Fawaz

**Comment by diaspj** [comment]
Dear Fawaz,

I think you didn't understand the files that I have sent to you.

The file "BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.gfa" that you have used to feed into BubbleGun using the command that you mention

"BubbleGun -g BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.gfa bchains --bubble_json Lactococcus_lactis_PANGENOME_Anchorwave.json --fasta Lactococcus_lactis_PANGENOME_Anchorwave.fasta"


is the file that I have got as output from BubbleGun itself (using the command "--chains_gfa") 

BubbleGun -g Lactococcus_lactis_PANGENOME_Anchorwave.gfa bchains --chains_gfa BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.gfa



So, you have used as input gfa file for BubbleGun the output of BubbleGun itself, "BubbleGun_Lactococcus_lactis_PANGENOME_Anchorwave.gfa", a file with a size of 515.0 kB (which I have identified with the prefix "BubbleGun_"). A file of 515.0 kB can't be a pangenome!


Since you are feeding BubbleGun with the gfa that BubbleGun itself outputs, I think that is not a good manner to check whether your fix to the software is working properly!


The pangenome gfa file that I am interested on genotyping is the file "Lactococcus_lactis_PANGENOME_Anchorwave.gfa", a file with a size of 285,8 MB (shown in the command "--chains_gfa" above).

I have sent this file to you from a gmail account, because my IST e-mail account does not allows sending files of MB size.

Please check your e-mail inbox for a second e-mail from an account pjdias00@gmail.com, which contains a google drive link that makes available the pangenome gfa.  If it is not there because of spam filters, I can sent it to you again by another mean that is more convenient for you.

Summing up, the file "Lactococcus_lactis_PANGENOME_Anchorwave.gfa", a file with a size of 285,8 MB, is the pangenome gfa that you should use to check whether your correction to the BubbleGun code is working properly or not (and the VCF file in attachment can be used to compare your BubbleGun fasta output with vg deconstruct, that comprises the expected nucleotide sequence)


Hope that this message was clear to you and helps you understand the input file I am using and the BubbleGun output files the software have outputted,


Best regards,

Paulo Dias


PS: I already installed the last BubbleGun version using the commands you have suggested, many thanks!

**Comment by fawaz-dabbaghieh** [comment]
Dear Paulo,

I understand the files you sent me. I tested on the smaller graph because the bigger graph was taking too long to finish and didn't want to keep waiting.
Moreover, looking at the bigger graph you sent, I see that nodes 2 and 3 have a huge number of edges (991115 edges involving node 2 and 366167 involving node 3), I am surprised you were even able to run BubbleGun on such a graph in the first place. How long did it take to run BubbleGun on the big graph?

Anyway, the issue here is between the output JSON file and the FASTA being outputted by BubbleGun, regardless of whether the graph I am testing on is a pangenome or not, this should not affect the assessment of consistency between the two files, as BubbleGun is agnostic to where the GFA came from. Therefore, it doesn't matter if the GFA I am using is a GFA that was previously outputted by BubbleGun or not.

Anyway, I understand your concerns, another reason I ran on the smaller graph is to be able to assess everything visually with Bandage, and after looking at Bandage, the JSON file, the GFA file, and the FASTA, everything was consistent for me after pushing the last commit. Moreover, I have tested it on 3 different graphs of my own and they were all consistent.

I am not sure how many more tests I can still do, as for me, after the last commit, it's all consistent and the bug seems to be fixed.

Please make sure to completely remove BubbleGun before reinstalling, as I had this problem before where during the installation of an updated commit, BubbleGun was not properly replaced.

Best,
Fawaz

**Comment by diaspj** [comment]
Dear Fawaz,

A) I am running the fixed version of BubbleGun right now on the Pangenome of Lactococcus_lactis, to confirm that your fix to the command "BubbleGun bchains --fasta" is working properly.

When Bubblegun finishes calculating the bubbles on the Pangenome of Lactococcus_lactis, I will report back in this issue confirming the sucess of the fix.

B) Regarding the performance of BubbleGun on working on big graphs, I leave two examples exemplifying the robustness of your software when analyzing two Pangenomes which presented the more extreme cases, one regarding the highest number of bubbles I have encounter (first example) and another regarding the longest running time until finishing the identification of all bubbles (second example).

As you can see from the two examples bellow, BubbleGun is working quite well on big graphs!


Example 1:

In Saccharomyces uvarum (summary bellow), comprising a pangenome composed of 5 genome sequences, BubbleGun runned for about 1 day to finish the calculation of all bubbles, totalling more than half a million bubbles.

BubbleGun -g Saccharomyces_uvarum_PANGENOME_Anchorwave.gfa bchains --bubble_json BubbleGun_Saccharomyces_uvarum_PANGENOME_Anchorwave.json

The number of Simple Bubbles is 531932
The number of Superbubbles is 1200
The number of insertions is 41676
Sequence coverage of the bubble chains is 83.71066355543941%
Node coverage of the bubble chains is 75.99556346992829%


Example 2:

In another microorganism, Burkholderia cepacia (summary bellow), comprising a pangenome composed of 16 genome sequences, BubbleGun runned for about 3 weeks to finish the calculation of all bubbles, totalling more than 70000 bubbles.

BubbleGun -g Burkholderia_cepacia_PANGENOME_Anchorwave.gfa bchains --bubble_json BubbleGun_Burkholderia_cepacia_PANGENOME_Anchorwave.json

The number of Simple Bubbles is 67801
The number of Superbubbles is 7881
The number of insertions is 5884
Sequence coverage of the bubble chains is 5.764042793641741%
Node coverage of the bubble chains is 7.97081683196567%
The longest chain seq-wise has 26098 bp
The longest chain bubble_wise has 1619 bubbles

Moreover, BubbleGun never crashed during all computations I have performed over more than 15 microorganisms (bacteria and yeast), although the RAM memory I have on my workstation is only 64Gb.


C) Finally, I leave a question regarding the BubbleGun fasta output files:

It would be possible for you to include a change in the code of BubbleGUn to allow to output the fasta files of the bubbles tagged with "super" and "insertion" and not only to the "simple" tagged ones?  

In case this new ability is a quick and easy fix to the BubbleGun code, I would like to suggest that it should be added to the software since it would allow a more quick confirmation of the DNA sequences associated with each variant present in pangenomes objects. 

In case it is a difficult fix, I am submitting another issue regarding the ability of BubbleGun ouputting gfa files in an alternative way, maybe these can be more easily implemented...


Best,

Paulo Dias

**Comment by fawaz-dabbaghieh** [comment]
Dear Paulo,

These runtimes are very weird, considering that I deal with much much bigger graphs than the ones you are working with and BubbleGun usually finishes in less than half an hour (I recently ran it on human pangenome chromosome 1, with almost 6 million nodes and it took few minutes to run completely reporting way more bubbles than in your graphs).

I am assuming these extremely long run times are happening because of these "weird" nodes that have a huge amount of edges coming through them, which I am guessing are because of VG, as I had a similar problem with VG years ago. I am not sure how these nodes come to be, and what their purpose is. But they are causing the BubbleGun algorithm to take much longer than it's supposed to be as every time the algorithm starts from a node, it ends up entering one of these "problematic" nodes, and then the algorithm needs to check every neighbor of these nodes, before it reports a bubble it found or aborts.

Have you looked into other methods to build your pangenome? For example, using minigraph, and their incremental graph generator method ([Here](https://github.com/lh3/minigraph)). Usually, `minigraph` keeps the topology of the graph linear.
Or using a `minigraph` cactus graph similar to the HPRC method  ([Here](https://github.com/human-pangenomics/hpp_pangenome_resources)). Minigraph-cactus graphs are much bigger than the ones produced with only minigraph, but BubbleGun should be able to handle these graphs very easily.

As for the last point, I can edit the code to output the insertions. However, superbubbles are not possible to output and I don't think I will be adding that to BubbleGun. The reason is that a superbubble can be arbitrarily big, therefore, outputting all possible paths connecting the source and sink node can easily explode in number, especially since in a superbubble, the paths between the source and the sink nodes do not have to be disjoint, which also adds to the complexity of calculating every possible path between the source and sink.

Best,
Fawaz

**Comment by diaspj** [comment]
Dear Fawaz,

1) Maybe the runtimes I am reporting for B. cepacia (3 weeks) derive from the fact that I was using an outdated version of BubbleGun. 

However, in the overwhelming majority of the cases, BubbleGun only takes a 1-2 days to conclude the calculation of all bubbles, sometimes a few more days in the Pangenome graphs made using the approach I describe bellow.

I remember that I have installed at some point in time your "development" version of BubbleGun, version 1.1.0, that still remains in the github website, maybe it was this abandoned version that still dwells in the site that I was using when I started to use BubbleGun (I suggest this version should be removed from the site).

And maybe is a good thing to add to BubbleGun a "--version" option that outputs the current version of the software, and any modification to the BubbleGun code should be idenfied in the github as a "new version" so the code that is being used becomes identified inequivocally.

The BubbleGun version present in the pip site should also be updated, it comprised a version identified as 1.7.0 which I also installed.


2) I am using seqwish to induce the pangenome graph using the alignment algorithm made available by Anchorwave, so I think is not VG but seqwish that is originating those nodes with many edges.

I think that using only Minigraph is not a good way to obtain the Pangenome graph, since the alignment is not good at the nucleotide level. Presently, I am also using scripting to combine the Minigraph-cactus as the alignment algorithm with seqwish, so that the resulting file is an unbiased graph.

This will allow to analyze the effect of alignment algorithm, Anchorwave versus Minigraph-cactus, by comparing the set of genotypes that are obtained by the two methodologies, both using seqwish to induce the graph.

I am not worried about seqwish originating nodes with many edges because, at the level of microorganisms (bacteria and yeast), BubbleGun handles that task without crashing and always generate the corresponding json and gfa files.

I have only observed the crashing of BubbleGun when using the command "BubbleGun compact" on a couple of species (as you mention, probably is because of these highly connected nodes). If you are thinking developing further this software, this is the only algorithm of BubbleGun that I have noticed that have shown a "tendency" to crash, so maybe this algorithm have room to be improved so that this graph operation can become more robust. 


3) Regarding the last point, my opinion is that adding the ability to BubbleGun of outputting the insertions to the fasta file will be very useful, at least, from my perspective and of the work I am developing based on the  json output file.

Presently, I am developing code to add genotypes based on the BubbleGun json output file, so it useful to have a way to check whether the sequences in my output files match the ones in the output of BubbleGun. 

The fasta file generated by the command "BubbleGun bchains --fasta", in case it contains the nucleotides sequences of bubbles tagged as "simple", "insertion" and "super", would be a simple way of checking if the nucleotide sequences my script is outputting correspond to the nucleotide sequences BubbleGun is outputting, and even compare with the nucleotide sequences comprised in the vcf file generated by the "vg deconstruct" command.

I understand your comments regarding explosion of paths between the source and the sink nodes. 

However, operationally, Pangenome graphs are always built from a limited number of genome sequences.

This fact occurs because there is a limited number of genome sequences available and because the software used to built them are constrained to handle a limited number of genome sequences because of RAM limits and/or disk size limits.

So, your concerns regarding the explosion of paths does not occurs operationally, only can occur in extreme cases, with Pangenome graphs made up of hundreds or thousands of genome sequences, and/or in case of Pangenome graphs of complex eukaryotic genome sequences such as humans, comprising many different genomes.

In fact, what I have observed from the analysis of various Pangenome graphs from various microorganisms, the paths between the source and the sink nodes are normally disjoint (I don't want to use the word "always" because I did not check all cases in the vcf file).

Since you are not adding the ability to output bubbles tagged as "super", I propose an alternative methodology so that the BubbleGun allow an easy access to the nucleotide sequence comprised in the bubbles tagged as "super".

Since BubbleGun already have the ability of outputting a single gfa file which, I infer, results from the concatenation of the single gfa files comprising each chain identified by the software, I would like to ask if it is possible to add the ability to output each individual chains to a corresponding single gfa file. 

Moreover, in case it is an easy fix, maybe you could extend this ability and also allow to output each individual bubble to corresponding single gfa files.

These new abilities would allow an easy access the nucleotide sequence of bubbles tagged with "super", at the chain or at the bubble levels, and would allow to circumvent the non-availability of the nucleotide sequences corresponding to the "super" bubbles in the fasta file.


4) I have just obtained the .json, .gfa and .fasta files using the fixed version of BubbleGun, after I have the time to analyze these files (and I have already observed that the order of the bubbles in the new .json file is different from one comprised in the first .json file that I had obtained using the previous version of BubbleGun), I will report and confirm in this issue whether the fix is working properly.


Best,

Paulo Dias

**Comment by fawaz-dabbaghieh** [comment]
Dear Paulo,

I don't think the recent runtimes are connected to the old or new version, but these extremely long runtimes are happening because of these hyperconnected nodes in the graph, and at the moment, I cannot think of a way to avoid this, as the algorithm needs to check all these possible edges. And when BubbleGun was written, such an extreme case was not considered.

For BubbleGun compact, I'll need to look into why it might crash with such graphs. the `compact` subcommand is designed to combine linear paths (i.e. a path of several nodes where each node has an in and out-degree of 1), so maybe the graphs you have with these hyperconnected nodes are causing some confusion in the algorithm, but as I don't know what was causing the crash to begin with, it'll be hard to debug for now, if you get this again, please report it as an issue.

When it comes to versioning, usually one does not remove a previous version when a new version is committed to Github, that's the point of having a GitHub repository, the new versions are pushed as commits to the repo, and one needs to "pull" the new version and re-install it (Which is not ideal, I know). Of course, it's better when this is all done through Pypi. However, I usually don't add the minor versions to Pypi immediately, but wait until it's a major version change. However, I will add the new developments to Pypi soon.

About superbubbles, as BubbleGun is a general-purpose tool, I cannot only assume your use case is the default one and that superbubbles will have a small number of "disjoint" paths. I also disagree that in practice this won't happen, I, for example, work with other pangenome graphs that have very big Superbubbles (e.g. in the thousands of nodes for one superbubble). At the moment, there's no straightforward or agreed-upon way to deconstruct a superbubble into individual sequences. Therefore, I will refrain at the moment from outputting such information in the FASTA output.

As for your suggestion about outputting every chain in a separate GFA file, I will not do that as this again can explode dramatically for big graphs, and can easily end up outputting 100s of thousands of files which can overwhelm someone's computer or system. However, this is already implemented in BubbleGun, to output a specific chain or chains you're interested in, check the `chainout` subcommand, it takes the graph, the JSON file, and a list of chain IDs and outputs these chains into a separate GFA file.

Moreover, BubbleGun can be used as a library, i.e. you can write your own scripts in Python that use BubbleGun's internal classes. For example, you can easily load a GFA, output a GFA, or output a subgraph of the GFA by giving a list of the nodes you're interested in outputting. You can look at the Graph.py and Node.py classes to check what kind of functionality is already implemented.

Best,
Fawaz

**Comment by fawaz-dabbaghieh** [comment]
As for the version, I added a `--version` command to output the installed version
